syncd/Syncd.cpp

```diff
@@ -802,6 +802,67 @@ sai_status_t Syncd::processBulkEntry(
 
     sai_status_t all = SAI_STATUS_SUCCESS;
 
+#ifdef SAI_SUPPORT_BULK_API
+    std::vector<sai_object_meta_key_t> metaKeys(objectIds.size());
+    std::vector<uint32_t> attr_counts(objectIds.size());
+
+    std::vector<const sai_attribute_t*> attr_lists(objectIds.size());
+    for (size_t idx = 0; idx < objectIds.size(); idx++)
+    {
+        metaKeys[idx].objecttype = objectType;
+
+        switch (objectType)
+        {
+            case SAI_OBJECT_TYPE_ROUTE_ENTRY:
+                sai_deserialize_route_entry(objectIds[idx], metaKeys[idx].objectkey.key.route_entry);
+                break;
+
+            case SAI_OBJECT_TYPE_FDB_ENTRY:
+                sai_deserialize_fdb_entry(objectIds[idx], metaKeys[idx].objectkey.key.fdb_entry);
+                break;
+
+            default:
+                SWSS_LOG_THROW("object %s not implemented, FIXME", sai_serialize_object_type(objectType).c_str());
+        }
+        m_translator->translateVidToRid(metaKeys[idx]);
+
+        attr_counts[idx] = attributes[idx]->get_attr_count();
+        attr_lists[idx] = attributes[idx]->get_attr_list();
+    }
+
+    sai_bulk_op_error_mode_t mode = SAI_BULK_OP_ERROR_MODE_IGNORE_ERROR;
+
+    switch (api)
+    {
+        case SAI_COMMON_API_BULK_CREATE:
+            all = m_vendorSai->bulkCreate(metaKeys.data(), SAI_NULL_OBJECT_ID, (uint32_t)objectIds.size(), attr_counts.data(),
+                                            attr_lists.data(), mode, statuses.data());
+            break;
+
+        case SAI_COMMON_API_BULK_REMOVE:
+            all = m_vendorSai->bulkRemove(metaKeys.data(), (uint32_t)objectIds.size(), mode, statuses.data());
+            break;
+
+        case SAI_COMMON_API_BULK_SET:
+            all = m_vendorSai->bulkSet(metaKeys.data(), (uint32_t)objectIds.size(),
+                                             attr_lists.data(), mode, statuses.data());
+            break;
+
+        default:
+            all = SAI_STATUS_NOT_SUPPORTED;
+            SWSS_LOG_THROW("api %d is not supported in bulk mode", api);
+    }
+
+    if (all != SAI_STATUS_NOT_SUPPORTED && all != SAI_STATUS_NOT_IMPLEMENTED)
```

Comment:
But what if bulk functions is not implemented in SAI? then we skip sendResponse in 858 line and try to process this entries in loop from 866 line.
Also if bulk functions is supported and implemented, then we need send response in 858 line and exit from function in 862 line.
In both case response will be send in 858 line (if bulk functions are supported) or in 925 line (if bulk functions are not supported)
This check was added in case if bulk functions is not implemented and this case should not affect flow.

Reply:
SAI_STATUS_NOT_SUPPORTED or SAI_STATUS_NOT_IMPLEMENTED status can be returned from bulkCreate/bulkSet/bulkRemove functions

Reply:
that behavior depends whether we want to fallback to loop 1by1 if vendor bulk api is not available, this was not mendioned